# <b>🚕 Urban Mobility Intelligence</b>

## <b>Phase 1 — Data Acquisition & Understanding</b>

### <b>Dataset</b>
NYC TLC Yellow Taxi Trip Record Data

### <b>Current Dataset</b>
January 2025

### <b>Objective</b>

The objective of this notebook is to understand the structure,
size, schema, data types, time range, completeness, uniqueness,
and basic statistical characteristics of the NYC Yellow Taxi
January 2025 dataset.

This analysis will be performed before data cleaning and
exploratory data analysis.

### <b>Business Context</b>

The dataset contains real-world NYC Yellow Taxi trip records.
Understanding the data correctly is essential before building
models for:

1. Taxi Fare Prediction
2. Taxi Demand Prediction
3. Taxi Demand Forecasting

#### <b> STEP 1: IMPORT THE LIBRARIES</b>

In [1]:
import pandas as pd
import numpy as np
from pathlib import Path

#### <b>STEP 2: VERIFY THE LOCATION</b>

In [2]:
from pathlib import Path

project_root = Path.cwd().parent

print("Current notebook directory:")
print(Path.cwd())

print("\nProject root:")
print(project_root)

Current notebook directory:
d:\MachineLearningProjects\Urban-Mobility-Intelligence\notebooks

Project root:
d:\MachineLearningProjects\Urban-Mobility-Intelligence


#### <b>STEP 3: DEFINE RAW DATA PATH</b>

In [3]:
raw_data_path = project_root / "data" / "raw"

print(raw_data_path)

d:\MachineLearningProjects\Urban-Mobility-Intelligence\data\raw


In [4]:
dataset_path = raw_data_path / "yellow_tripdata_2025-01.parquet"

print(dataset_path)

d:\MachineLearningProjects\Urban-Mobility-Intelligence\data\raw\yellow_tripdata_2025-01.parquet


#### <b>STEP 4: VERIFY FILE EXISTENCE</b>

In [5]:
print("Dataset exists:", dataset_path.exists())

Dataset exists: True


#### <b>STEP 5: CHECK FILE SIZE</b>

In [6]:
file_size_bytes = dataset_path.stat().st_size
file_size_gb = file_size_bytes / (1024 ** 3)

print(f"File size: {file_size_gb:.2f} GB")

File size: 0.06 GB


#### <b>STEP 6: INSPECT PARQUET METADATA</b>

In [7]:
import pyarrow.parquet as pq

parquet_file = pq.ParquetFile(dataset_path)

print("Number of rows:", parquet_file.metadata.num_rows)
print("Number of columns:", parquet_file.metadata.num_columns)
print("Number of row groups:", parquet_file.num_row_groups)

Number of rows: 3475226
Number of columns: 20
Number of row groups: 4


In [8]:
parquet_file.schema

required group field_id=-1 schema {
  optional int32 field_id=-1 VendorID;
  optional int64 field_id=-1 tpep_pickup_datetime (Timestamp(isAdjustedToUTC=false, timeUnit=microseconds, is_from_converted_type=false, force_set_converted_type=false));
  optional int64 field_id=-1 tpep_dropoff_datetime (Timestamp(isAdjustedToUTC=false, timeUnit=microseconds, is_from_converted_type=false, force_set_converted_type=false));
  optional int64 field_id=-1 passenger_count;
  optional double field_id=-1 trip_distance;
  optional int64 field_id=-1 RatecodeID;
  optional binary field_id=-1 store_and_fwd_flag (String);
  optional int32 field_id=-1 PULocationID;
  optional int32 field_id=-1 DOLocationID;
  optional int64 field_id=-1 payment_type;
  optional double field_id=-1 fare_amount;
  optional double field_id=-1 extra;
  optional double field_id=-1 mta_tax;
  optional double field_id=-1 tip_amount;
  optional double field_id=-1 tolls_amount;
  optional double field_id=-1 improvement_surcharge;
  op

#### <b>STEP 7: INSPECT COLUMN NAMES</b>

In [9]:
columns = parquet_file.schema_arrow.names

print("Number of columns:", len(columns))
print("\nColumns:")

for i, column in enumerate(columns, start=1):
    print(f"{i}. {column}")

Number of columns: 20

Columns:
1. VendorID
2. tpep_pickup_datetime
3. tpep_dropoff_datetime
4. passenger_count
5. trip_distance
6. RatecodeID
7. store_and_fwd_flag
8. PULocationID
9. DOLocationID
10. payment_type
11. fare_amount
12. extra
13. mta_tax
14. tip_amount
15. tolls_amount
16. improvement_surcharge
17. total_amount
18. congestion_surcharge
19. Airport_fee
20. cbd_congestion_fee


#### <b>STEP 8: LOAD THE DATASET</b>

In [10]:
df = pd.read_parquet(dataset_path)

print("Dataset loaded successfully.")
print(f"Rows: {df.shape[0]:,}")
print(f"Columns: {df.shape[1]}")

Dataset loaded successfully.
Rows: 3,475,226
Columns: 20


#### <b>STEP 9: DATAFRAME MEMORY USAGE</b>

In [11]:
memory_usage_mb = df.memory_usage(deep=True).sum() / (1024 ** 2)

print(f"DataFrame memory usage: {memory_usage_mb:.2f} MB")

DataFrame memory usage: 493.47 MB


#### <b>STEP 10: INSPECT DATA</b>

In [12]:
df.head()

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee
0,1,2025-01-01 00:18:38,2025-01-01 00:26:59,1.0,1.60,1.0,N,229,237,1,10.0,3.5,0.5,3.00,0.0,1.0,18.00,2.5,0.0,0.0
1,1,2025-01-01 00:32:40,2025-01-01 00:35:13,1.0,0.50,1.0,N,236,237,1,5.1,3.5,0.5,2.02,0.0,1.0,12.12,2.5,0.0,0.0
2,1,2025-01-01 00:44:04,2025-01-01 00:46:01,1.0,0.60,1.0,N,141,141,1,5.1,3.5,0.5,2.00,0.0,1.0,12.10,2.5,0.0,0.0
3,2,2025-01-01 00:14:27,2025-01-01 00:20:01,3.0,0.52,1.0,N,244,244,2,7.2,1.0,0.5,0.00,0.0,1.0,9.70,0.0,0.0,0.0
4,2,2025-01-01 00:21:34,2025-01-01 00:25:06,3.0,0.66,1.0,N,244,116,2,5.8,1.0,0.5,0.00,0.0,1.0,8.30,0.0,0.0,0.0


In [13]:
df.tail()

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee
3475221,2,2025-01-31 23:01:48,2025-01-31 23:16:29,NaN,3.35,NaN,NaN,79,237,0,15.85,0.0,0.5,0.0,0.0,1.0,20.60,NaN,NaN,0.75
3475222,2,2025-01-31 23:50:29,2025-02-01 00:17:27,NaN,8.73,NaN,NaN,161,116,0,28.14,0.0,0.5,0.0,0.0,1.0,32.89,NaN,NaN,0.75
3475223,2,2025-01-31 23:26:59,2025-01-31 23:43:01,NaN,2.64,NaN,NaN,144,246,0,14.91,0.0,0.5,0.0,0.0,1.0,19.66,NaN,NaN,0.75
3475224,2,2025-01-31 23:14:34,2025-01-31 23:34:52,NaN,3.16,NaN,NaN,142,107,0,17.55,0.0,0.5,0.0,0.0,1.0,22.30,NaN,NaN,0.75
3475225,2,2025-01-31 23:56:42,2025-02-01 00:07:27,NaN,2.29,NaN,NaN,237,238,0,12.09,0.0,0.5,0.0,0.0,1.0,16.09,NaN,NaN,0.00


#### <b>STEP 11: CHECK SHAPE</b>

In [14]:
print("Rows:", f"{df.shape[0]:,}")
print("Columns:", df.shape[1])

Rows: 3,475,226
Columns: 20


#### <b>STEP 12: CHECK DATATYPES</b>

In [15]:
df.dtypes

VendorID                          int32
tpep_pickup_datetime     datetime64[us]
tpep_dropoff_datetime    datetime64[us]
passenger_count                 float64
trip_distance                   float64
RatecodeID                      float64
store_and_fwd_flag                  str
PULocationID                      int32
DOLocationID                      int32
payment_type                      int64
fare_amount                     float64
extra                           float64
mta_tax                         float64
tip_amount                      float64
tolls_amount                    float64
improvement_surcharge           float64
total_amount                    float64
congestion_surcharge            float64
Airport_fee                     float64
cbd_congestion_fee              float64
dtype: object

#### <b>STEP 13: CHECK DATASET INFORMATION</b>

In [16]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 3475226 entries, 0 to 3475225
Data columns (total 20 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int32         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     str           
 7   PULocationID           int32         
 8   DOLocationID           int32         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  Airport_fee            float64   

#### <b>STEP 14: CHECK MISSING VALUES</b>

##### <b>Missing Value Analysis</b>

Before performing data cleaning, we need to determine whether
the raw dataset contains missing values in any of its columns.

At this stage, missing values will only be measured and documented.
No records will be removed or modified.

In [17]:
missing_values = df.isnull().sum()

missing_summary = pd.DataFrame({
    "missing_count": missing_values,
    "missing_percentage": (missing_values / len(df)) * 100
})

missing_summary = missing_summary.sort_values(
    by="missing_count",
    ascending=False
)

missing_summary

,missing_count,missing_percentage
passenger_count,540149,15.542845
congestion_surcharge,540149,15.542845
store_and_fwd_flag,540149,15.542845
RatecodeID,540149,15.542845
Airport_fee,540149,15.542845
tpep_dropoff_datetime,0,0.000000
VendorID,0,0.000000
tpep_pickup_datetime,0,0.000000
DOLocationID,0,0.000000
payment_type,0,0.000000


#### <b>STEP 15: CHECK MISSING VALUE COLUMNS</b>

In [18]:
missing_columns = missing_summary[
    missing_summary["missing_count"] > 0
]

missing_columns

,missing_count,missing_percentage
passenger_count,540149,15.542845
congestion_surcharge,540149,15.542845
store_and_fwd_flag,540149,15.542845
RatecodeID,540149,15.542845
Airport_fee,540149,15.542845


#### <b>STEP 16: CHECK DUPLICATES</b>

##### <b>Duplicate Record Analysis</b>

Duplicate records can affect demand calculations, fare analysis,
and machine-learning models.

We will measure duplicate records before deciding whether any
records should be removed.

In [19]:
duplicate_count = df.duplicated().sum()

print(f"Duplicate records: {duplicate_count:,}")
print(f"Duplicate percentage: {(duplicate_count / len(df)) * 100:.4f}%")

Duplicate records: 0
Duplicate percentage: 0.0000%


#### <b>STEP 17: CHECK TEMPORAL RANGE</b>

In [20]:
print("Pickup datetime range:")
print("Start:", df["tpep_pickup_datetime"].min())
print("End:", df["tpep_pickup_datetime"].max())

print("\nDrop-off datetime range:")
print("Start:", df["tpep_dropoff_datetime"].min())
print("End:", df["tpep_dropoff_datetime"].max())

Pickup datetime range:
Start: 2024-12-31 20:47:55
End: 2025-02-01 00:00:44

Drop-off datetime range:
Start: 2024-12-18 07:52:40
End: 2025-02-01 23:44:11


#### <b>STEP 18: CHECK UNIQUE VALUES</b>

In [21]:
categorical_columns = [
    "VendorID",
    "RatecodeID",
    "store_and_fwd_flag",
    "payment_type"
]

for column in categorical_columns:
    print(f"\n{'=' * 60}")
    print(f"{column}")
    print(f"{'=' * 60}")
    print(df[column].value_counts(dropna=False))


VendorID
VendorID
2    2719860
1     753671
7       1206
6        489
Name: count, dtype: int64

RatecodeID
RatecodeID
1.0     2756472
NaN      540149
2.0       94420
99.0      41963
5.0       26501
3.0        8622
4.0        7092
6.0           7
Name: count, dtype: int64

store_and_fwd_flag
store_and_fwd_flag
N      2927431
NaN     540149
Y         7646
Name: count, dtype: int64

payment_type
payment_type
1    2444393
0     540149
2     390429
4      76481
3      23773
5          1
Name: count, dtype: int64


#### <b>STEP 19: INSPECT LOACTION IDs</b>

In [22]:
print("Unique pickup locations:", df["PULocationID"].nunique())
print("Unique drop-off locations:", df["DOLocationID"].nunique())

Unique pickup locations: 261
Unique drop-off locations: 260


In [23]:
print("Pickup location range:")
print(df["PULocationID"].min(), "to", df["PULocationID"].max())

print("\nDrop-off location range:")
print(df["DOLocationID"].min(), "to", df["DOLocationID"].max())

Pickup location range:
1 to 265

Drop-off location range:
1 to 265


#### <b>STEP 20: Shared Missing-Value Pattern</b>

Several columns contain exactly the same number of missing values.
Before deciding how to handle these missing values, we need to
determine whether the missing values occur in the same records.

This will help us understand whether the missingness is associated
with a particular subset of trips.

In [24]:
shared_missing_columns = [
    "passenger_count",
    "congestion_surcharge",
    "store_and_fwd_flag",
    "RatecodeID",
    "Airport_fee"
]

shared_missing_mask = df[shared_missing_columns].isnull().all(axis=1)

print(
    "Records missing all five fields:",
    f"{shared_missing_mask.sum():,}"
)

print(
    "Percentage:",
    f"{shared_missing_mask.mean() * 100:.2f}%"
)

Records missing all five fields: 540,149
Percentage: 15.54%


In [25]:
partial_missing_mask = (
    df[shared_missing_columns].isnull().sum(axis=1).between(1, 4)
)

print(
    "Records with partial missingness:",
    f"{partial_missing_mask.sum():,}"
)

Records with partial missingness: 0


#### <b>STEP 21: RECORDS WITH SHARED MISSING-VALUES</b>

In [26]:
df.loc[
    shared_missing_mask,
    [
        "VendorID",
        "tpep_pickup_datetime",
        "tpep_dropoff_datetime",
        "PULocationID",
        "DOLocationID",
        "payment_type",
        "fare_amount",
        "total_amount"
    ]
].head(10)

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,PULocationID,DOLocationID,payment_type,fare_amount,total_amount
2935077,2,2025-01-01 00:39:17,2025-01-01 01:10:42,48,42,0,2.32,14.90
2935078,2,2025-01-01 00:33:40,2025-01-01 01:10:34,4,48,0,6.02,10.02
2935079,1,2025-01-01 00:09:32,2025-01-01 00:12:38,158,158,0,22.72,26.72
2935080,2,2025-01-01 00:21:42,2025-01-01 00:36:29,137,211,0,-2.35,5.80
2935081,2,2025-01-01 00:56:04,2025-01-01 01:06:11,140,233,0,5.06,9.06
2935082,1,2025-01-01 00:33:36,2025-01-01 00:54:48,263,162,0,18.42,22.42
2935083,2,2025-01-01 00:50:25,2025-01-01 00:57:34,142,239,0,14.16,20.16
2935084,2,2025-01-01 00:52:42,2025-01-01 01:27:35,144,263,0,23.13,27.13
2935085,2,2025-01-01 00:41:07,2025-01-01 01:02:39,50,107,0,21.83,26.83
2935086,1,2025-01-01 00:55:15,2025-01-01 01:43:25,107,50,0,27.55,31.55


#### <b>STEP 22: SHARED MISSING VALUES WITH VENDORID</b>

In [27]:
df.loc[shared_missing_mask, "VendorID"].value_counts()

VendorID
2    451456
1     88204
6       489
Name: count, dtype: int64

In [28]:
df.loc[shared_missing_mask, "payment_type"].value_counts()

payment_type
0    540149
Name: count, dtype: int64

#### <b>STEP 23: BOUNDARY DATES INVESTIGATION</b>

In [29]:
january_start = pd.Timestamp("2025-01-01")
february_start = pd.Timestamp("2025-02-01")

outside_january = (
    (df["tpep_pickup_datetime"] < january_start) |
    (df["tpep_pickup_datetime"] >= february_start)
)

print(
    "Pickup records outside January 2025:",
    f"{outside_january.sum():,}"
)

print(
    "Percentage:",
    f"{outside_january.mean() * 100:.4f}%"
)

Pickup records outside January 2025: 22
Percentage: 0.0006%


In [30]:
df.loc[
    outside_january,
    [
        "tpep_pickup_datetime",
        "tpep_dropoff_datetime",
        "VendorID",
        "PULocationID",
        "DOLocationID",
        "fare_amount",
        "total_amount"
    ]
].head(10)

,tpep_pickup_datetime,tpep_dropoff_datetime,VendorID,PULocationID,DOLocationID,fare_amount,total_amount
605,2024-12-31 23:30:03,2024-12-31 23:43:02,2,246,13,16.3,26.62
687,2024-12-31 23:31:38,2024-12-31 23:41:48,2,43,140,10.7,15.70
688,2024-12-31 23:46:38,2025-01-01 00:03:03,2,229,24,19.1,24.10
861,2024-12-31 23:56:19,2025-01-01 00:11:19,2,68,107,14.9,23.88
1108,2024-12-31 23:55:37,2025-01-01 00:01:26,2,56,56,7.9,10.40
1312,2024-12-31 23:52:40,2025-01-01 00:23:03,2,114,151,34.5,47.40
2276,2024-12-31 23:49:24,2024-12-31 23:57:30,2,142,151,12.1,20.52
3941,2024-12-31 23:27:13,2024-12-31 23:35:48,2,170,141,10.0,18.00
3942,2024-12-31 23:37:42,2024-12-31 23:43:10,2,229,141,7.2,14.03
4324,2024-12-31 23:51:20,2025-01-01 00:00:00,2,138,40,48.5,69.30


In [31]:
outside_january_dropoff = (
    (df["tpep_dropoff_datetime"] < january_start) |
    (df["tpep_dropoff_datetime"] >= february_start)
)

print(
    "Drop-off records outside January 2025:",
    f"{outside_january_dropoff.sum():,}"
)

print(
    "Percentage:",
    f"{outside_january_dropoff.mean() * 100:.4f}%"
)

Drop-off records outside January 2025: 1,811
Percentage: 0.0521%


#### <b>STEP 24: NUMERICAL DATA UNDERSTANDING</b>

##### <b>Numerical Variable Analysis</b>

The next step is to understand the basic behavior of the major
numerical variables before defining data-cleaning rules.

We will examine:

- Passenger count
- Trip distance
- Fare amount
- Tip amount
- Tolls
- Total amount

At this stage, we will identify unusual values but will not
remove or modify any records.

In [32]:
numerical_columns = [
    "passenger_count",
    "trip_distance",
    "fare_amount",
    "tip_amount",
    "tolls_amount",
    "total_amount"
]

df[numerical_columns].describe().T

,count,mean,std,min,25%,50%,75%,max
passenger_count,2935077.0,1.297859,0.750750,0.00,1.00,1.00,1.00,9.00
trip_distance,3475226.0,5.855126,564.601600,0.00,0.98,1.67,3.10,276423.57
fare_amount,3475226.0,17.081803,463.472918,-900.00,8.60,12.11,19.50,863372.12
tip_amount,3475226.0,2.959813,3.779681,-86.00,0.00,2.45,3.93,400.00
tolls_amount,3475226.0,0.449308,2.002582,-126.94,0.00,0.00,0.00,170.94
total_amount,3475226.0,25.611292,463.658478,-901.00,15.20,19.95,27.78,863380.37


#### <b>STEP 25: COUNT ZERO AND NEGATIVE VALUES</b>

In [33]:
for column in numerical_columns:
    zero_count = (df[column] == 0).sum()
    negative_count = (df[column] < 0).sum()

    print(f"\n{column}")
    print(f"Zero values:     {zero_count:,}")
    print(f"Negative values: {negative_count:,}")


passenger_count
Zero values:     24,656
Negative values: 0

trip_distance
Zero values:     90,893
Negative values: 0

fare_amount
Zero values:     1,398
Negative values: 144,118

tip_amount
Zero values:     1,118,008
Negative values: 124

tolls_amount
Zero values:     3,259,590
Negative values: 4,559

total_amount
Zero values:     559
Negative values: 63,037


#### <b>STEP 26: TRIP DURATION</b>

In [34]:
trip_duration_minutes = (
    df["tpep_dropoff_datetime"] -
    df["tpep_pickup_datetime"]
).dt.total_seconds() / 60

trip_duration_minutes.describe()

count    3.475226e+06
mean     1.501812e+01
std      3.871358e+01
min     -5.147232e+04
25%      7.283333e+00
50%      1.170000e+01
75%      1.833333e+01
max      5.626317e+03
dtype: float64

##### <b>UNSUAL TRIP DURATION</b>

In [35]:
print(
    "Zero/negative duration:",
    f"{(trip_duration_minutes <= 0).sum():,}"
)

print(
    "Trips longer than 3 hours:",
    f"{(trip_duration_minutes > 180).sum():,}"
)

Zero/negative duration: 2,051
Trips longer than 3 hours: 1,377


##### <b>EXTREME TRIP-DISTANCE</b>

In [36]:
print("Zero-distance trips:")
print((df["trip_distance"] == 0).sum())

print("\nExtremely long trips (> 100 miles):")
print((df["trip_distance"] > 100).sum())

Zero-distance trips:
90893

Extremely long trips (> 100 miles):
162


#### <b>STEP 27: FARE CINSISTENCY</b>

In [37]:
fare_consistency = df[
    df["total_amount"] < df["fare_amount"]
]

print(
    "Records where total_amount < fare_amount:",
    f"{len(fare_consistency):,}"
)

Records where total_amount < fare_amount: 59,433


##### <b>EXTREME FARE</b>

In [38]:
print("Zero fare:")
print((df["fare_amount"] == 0).sum())

print("\nNegative fare:")
print((df["fare_amount"] < 0).sum())

print("\nFare greater than $500:")
print((df["fare_amount"] > 500).sum())

Zero fare:
1398

Negative fare:
144118

Fare greater than $500:
55


#### <b>Phase 1 — Data Understanding Findings</b>

##### <b>1. Dataset Overview</b>

The January 2025 NYC Yellow Taxi dataset contains 3,475,226
trip records and 20 columns.

The source file is stored in Parquet format and has a size of
approximately 0.06 GB. After loading into Pandas, the DataFrame
uses approximately 493.47 MB of memory.

The dataset contains trip timestamps, passenger information,
pickup and drop-off locations, trip distance, payment information,
fare components, and total trip charges.

---

##### <b>2. Data Completeness</b>

Five columns contain missing values:

- passenger_count
- RatecodeID
- store_and_fwd_flag
- congestion_surcharge
- Airport_fee

Each contains 540,149 missing values, representing approximately
15.54% of the dataset.

All five fields are missing together for the same records, and
there are no records with partial missingness across these fields.

The records with this shared missingness have payment_type = 0.

Missing records will not be removed at this stage. Their business
meaning will be investigated during the data-cleaning phase.

---

##### <b>3. Duplicate Records</b>

The dataset contains:

- 0 exact duplicate records
- 0.00% duplicate rate

Therefore, no duplicate-removal action is currently required.

---

##### <b>4. Temporal Coverage</b>

The pickup timestamps range from:

2024-12-31 20:47:55

to:

2025-02-01 00:00:44

Only 22 pickup records fall outside the January 2025 calendar
period.

Drop-off timestamps range from:

2024-12-18 07:52:40

to:

2025-02-01 23:44:11

There are 1,811 drop-off records outside the January 2025 period.

These records will be investigated before any filtering is applied,
because a January pickup may legitimately have a February drop-off.

---

##### <b>5. Location Coverage</b>

The dataset contains:

- 261 unique pickup locations
- 260 unique drop-off locations

Pickup and drop-off locations are represented using TLC Taxi Zone
identifiers.

These identifiers will later be connected to the official taxi
zone lookup data.

---

##### <b>6. Numerical Data Quality</b>

Several numerical variables contain potentially anomalous values.

`Passenger Count`

- 24,656 records have zero passengers.
- No negative passenger counts were observed.
- 540,149 records have missing passenger counts.

`Trip Distance`

- 90,893 records have zero distance.
- No negative distances were observed.
- Maximum distance is 276,423.57 miles.

The extremely high maximum indicates the presence of severe
outliers that require investigation.

`Fare Amount`

- 1,398 records have zero fare.
- 144,118 records have negative fares.
- Maximum fare is 863,372.12.

The extreme maximum and negative values require investigation
before model development.

`Tip Amount`

- 1,118,008 records have zero tips.
- 124 records have negative tips.
- Maximum tip is 400.

Zero tips are potentially valid, while negative tips require
investigation.

`Tolls Amount`

- 3,259,590 records have zero tolls.
- 4,559 records have negative toll values.
- Maximum toll amount is 170.94.

Zero tolls can be legitimate because not every trip uses toll roads.

`Total Amount`

- 559 records have zero total amount.
- 63,037 records have negative total amounts.
- Maximum total amount is 863,380.37.

The extreme values require investigation before the dataset is
used for modeling.

---

##### <b>7. Trip Duration</b>

Trip duration was derived temporarily from pickup and drop-off
timestamps for data-understanding purposes.

1,377 trips have durations greater than three hours.

These trips will be investigated during data cleaning rather than
automatically removed.

---

##### <b>8. Initial Business Observations</b>

The raw NYC taxi data contains several characteristics that can
materially affect business analysis and machine-learning models.

The most important observations are:

1. Taxi trips are highly concentrated around short distances,
   as indicated by the difference between mean and median trip
   distance.

2. Extreme trip-distance and fare values can heavily distort
   averages.

3. Missing values are structured and associated with
   payment_type = 0 rather than appearing randomly across
   individual columns.

4. The dataset contains a small number of trips crossing the
   calendar boundaries of the January dataset.

5. Taxi demand analysis will require aggregation by time and
   location.

6. Fare prediction will require careful treatment of anomalous
   financial and trip records.

7. Location IDs will need to be mapped to taxi-zone information
   before meaningful geographic analysis.

---

##### <b>9. Data-Quality Risks Identified</b>

The following risks will be investigated during Phase 2:

- Structured missing values
- Zero passenger counts
- Zero trip distances
- Extremely large trip distances
- Negative fares
- Extremely large fares
- Negative tips
- Negative tolls
- Negative total amounts
- Very long trips
- Calendar-boundary records
- Unusual rate codes
- Unusual vendor/payment codes

No records have been removed or modified during Phase 1.

---

#### <b> Phase 1 Conclusion</b>

The raw dataset is sufficiently rich and large enough to support
the project's three primary objectives:

1. Fare prediction
2. Taxi demand prediction
3. Taxi demand forecasting

However, the raw dataset contains several quality issues and
extreme values that must be investigated before exploratory
analysis and machine-learning development.

Therefore, the next phase will focus on systematic data-quality
investigation and cleaning.

---

**Phase 1 Status: COMPLETED**

**Next Phase: Phase 2 — Data Cleaning & Data Quality**